packages/legend-application-studio/src/stores/EditorStore.ts

```diff
@@ -1187,11 +894,73 @@ export class EditorStore implements CommandRegistrar {
     );
   }
 
-  changeLocalChangesState(): void {
-    if (this.isInFormMode) {
-      this.localChangesState = new FormLocalChangesState(this, this.sdlcState);
-    } else {
-      this.localChangesState = new TextLocalChangesState(this, this.sdlcState);
+  *switchModes(
+    to: GRAPH_EDITOR_MODE,
+    fallbackOptions?: {
+      isCompilationFailure?: boolean;
+      isGraphBuildFailure?: boolean;
+    },
+  ): GeneratorFn<void> {
+    switch (to) {
+      case GRAPH_EDITOR_MODE.GRAMMAR_TEXT: {
+        const graphEditorMode = new GraphEditSingleFileGrammarModeState(this);
+        try {
+          yield flowResult(
+            graphEditorMode.cleanupBeforeEntering(fallbackOptions),
```

Comment:
why are we still calling different set of methods when switching mode? I felt like we can also following a fix procedure, i.e. simplifying this down to bare-bone mode-independent logic, i.e.

```plaintext
attempt to leave and cleanup current mode
check if we can enter target mode
initialize target mode
switch to new mode
```